### PR TITLE
fix: Add proper unique constraints and remove soft deletion from models

### LINKED ISSUE #7093: Change unique constraints of soft deleted models
All models were changed to soft deletion without much thought about which ones make sense to be soft deleted and which ones don't. This leads to many errors as some models have unique constraints like discount code and when they are deleted and same code is added again, it throws an error. Thus, we need to revisit which models make sense to add soft deletion to, which should never be deleted (order, ticket, discount code) and which should not have soft deletion (custom forms)

This issue will focus on at least adding correct unique constraint, like (discount code, deleted at) instead of just discount code so that deleted models are not considered for uniqueness. And as we fix those, we'll remove soft deletion from some models and also add validations to prevent deletion in permanent records

### PR DESCRIPTION
**Pull request by Haider8**
Fixes #7093 

#### Short description of what this resolves:
Not all models requires soft deletion so we have to remove it from models like `CustomForms` and `EventSubTopic`. 

In some models (like `DiscountCode`) we need to add a unique constraint on `deleted_at` column.

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

File: app/models/discount_code.py [hunk 1]
```diff
@@ -13,7 +13,9 @@
 class DiscountCode(SoftDeletionModel):
     __tablename__ = "discount_codes"
     __table_args__ = (
-        UniqueConstraint('event_id', 'code', name='uq_event_discount_code'),
+        UniqueConstraint(
+            'event_id', 'code', 'deleted_at', name='uq_event_discount_code_deleted_at'
```
Comment (iamareebjamal) [inline app/models/discount_code.py:17]: Don't add deleted_at in constraint name, it's an implementation detail

Comment (Haider8) [reply app/models/discount_code.py:17]: Remove from all?

Comment (iamareebjamal) [reply app/models/discount_code.py:17]: Yes

**Comment by iamareebjamal** [comment]
Please fix the tests

**Comment by iamareebjamal** [comment]
Also, for all the models where deleted_at was removed, you need to delete the items with non null deleted_at before removing the column

**Comment by Haider8** [comment]
@iamareebjamal  There are 4 apib dredd fails like in `GET /v1/events/1/custom-form`. So, should I change that now?

**Comment by iamareebjamal** [comment]
Yes

**Comment by iamareebjamal** [comment]
> Also, for all the models where deleted_at was removed, you need to delete the items with non null deleted_at before removing the column

.

**Comment by Haider8** [comment]
@iamareebjamal Done

File: migrations/versions/rev-2020-06-30-20:07:55-f177c14710f1_.py [hunk 2]
```diff
@@ -18,6 +18,9 @@
 
 def upgrade():
     # ### commands auto generated by Alembic - please adjust! ###
+    op.execute('DELETE FROM custom_forms WHERE deleted_at IS NOT NULL')
+    op.execute('DELETE FROM event_sub_topics WHERE deleted_at IS NOT NULL')
```
Comment (iamareebjamal) [inline migrations/versions/rev-2020-06-30-20:07:55-f177c14710f1_.py:22]: Add equivalent statements in downgrade as well

**Comment by iamareebjamal** [comment]
Will merge after manual testing

File: app/models/ticket_fee.py [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
 
+from app.models.base import SoftDeletionModel
+
 from app.models import db
 
 DEFAULT_FEE = 0.0
 
 
-class TicketFees(db.Model):
+class TicketFees(SoftDeletionModel):
```
Comment (iamareebjamal) [inline app/models/ticket_fee.py:11]: Don't add more soft deletion

File: migrations/versions/rev-2020-06-30-20:07:55-f177c14710f1_.py (hunk 2 above)
Comment (Haider8) [reply migrations/versions/rev-2020-06-30-20:07:55-f177c14710f1_.py:22]: What will be the equivalent?

File: app/models/ticket_fee.py (hunk 3 above)
Comment (Haider8) [reply app/models/ticket_fee.py]: Removed :+1:

File: app/models/permission.py [hunk 4]
```diff
@@ -1,13 +1,17 @@
 from app.models import db
 
+from app.models.base import SoftDeletionModel
 
-class Permission(db.Model):
+
+class Permission(SoftDeletionModel):
```
Comment (iamareebjamal) [inline app/models/permission.py:6]: Here as well. Don't add more soft deletion. It has already complicated matters a lot

Comment (Haider8) [reply app/models/permission.py]: :+1:

**Pull request merged**